### PR TITLE
added docstring and comments to tree.py

### PR DESCRIPTION
**Pull request by skystar-p**
I added a bunch of docstrings and comments to my code.

Almost full-documented, so its code can be understood by just reading down.

### CODE REVIEW

File: backend/core/rule/tree.py [hunk 1]
```diff
@@ -31,67 +37,146 @@ class TreeLoaderException(Exception):
...
+            # If course is dict, this means it is not a leaf node, which can be recursively evaluated
             elif type(course) is dict:
+
+                # additional argument (maybe used by if_func)
                 args = course.get('args', [])
+
+                # hide falsy node or not
                 hide_false = course.get('hide_false', False)
+
+                # required credit for evaludated to True
```
Comment (sonowz) [inline backend/core/rule/tree.py]: required credit to be `evaluated` True

File: backend/core/rule/tree.py [hunk 2]
```diff
@@ -100,17 +185,46 @@ def load_tree(self, current_tree, previous_node):
...
                                             )
                     self.load_tree(course['courses'], current_node)
+
+                    # add this node
                     previous_node.add_children(current_node)
 
     def eval_tree(self, taken_list):
+        """Evaluate tree recursively
+
+        Constructor automately call this function, so it is not needed to be called explicitly.
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `automatically`

File: backend/core/rule/tree.py [hunk 3]
```diff
@@ -120,7 +234,31 @@ def find(self, name):
 
 class TreeNode(object):
     def __init__(self, data, properties, metadata, func=None, namespace=None, *args):
+        """Constructor of TreeNode
+
+        Args:
+            data: data of current node. more additional explain is in below
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `explanation`

File: backend/core/rule/tree.py [hunk 4]
```diff
@@ -120,7 +234,31 @@ def find(self, name):
...
+            properties: dict of property. will be deconstructed
+            metadata: data of any other information needed evaluating tree
+            func: evaluator function (closure)
```
Comment (sonowz) [inline backend/core/rule/tree.py]: um.. what does `closure` mean? `closed` function?

File: backend/core/rule/tree.py [hunk 5]
```diff
@@ -120,7 +234,31 @@ def find(self, name):
...
+            namespace: name which user actually see
+            *args: any other additional arguments
+        """
+
+        # self.data can contain various type of data.
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `types`

File: backend/core/rule/tree.py [hunk 6]
```diff
@@ -120,7 +234,31 @@ def find(self, name):
...
+        # If type of self.data is `None`, it means it is not
+        # yet evaluated. Maybe it is not leaf node, which
+        # means for being evaluate, all children node
+        # should be evaluated.
```
Comment (sonowz) [inline backend/core/rule/tree.py]: To clarify meaning,
```python
# means it is being evaluated, and all children node
# should be evaluated.
```

File: backend/core/rule/tree.py [hunk 7]
```diff
@@ -120,7 +234,31 @@ def find(self, name):
...
+        # If type of self.data is `Boolean`, it means it is
+        # evaluated.
```
Comment (sonowz) [inline backend/core/rule/tree.py]: To clarify meaning,
```python
# If type of self.data is `Boolean`, it means its
# evaluation is done.
```

File: backend/core/rule/tree.py [hunk 8]
```diff
@@ -129,16 +267,27 @@ def __init__(self, data, properties, metadata, func=None, namespace=None, *args)
         self.metadata = metadata
         self.func = func
         self.namespace = namespace
+
+        # maybe this variable needed for user-friendly interface
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `will be needed`

File: backend/core/rule/tree.py [hunk 9]
```diff
@@ -149,6 +298,15 @@ def __repr__(self):
         return '{0} <{1}>'.format(self.namespace, self.data)
 
     def tree_into_str(self, depth=0, hide=False):
+        """Convert tree to tree string
+
+        Args:
+            depth: current depth. this will be appeared by count of tabs
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `this will be applied to the number of tabs`?

File: backend/core/rule/tree.py [hunk 10]
```diff
@@ -149,6 +298,15 @@ def __repr__(self):
...
+            hide: boolean about hide falsy node or not
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `hiding`

File: backend/core/rule/tree.py [hunk 11]
```diff
@@ -165,6 +323,14 @@ def tree_into_str(self, depth=0, hide=False):
         return current_str
 
     def tree_into_dict(self, hide=False):
+        """Convert tree to tree dictionary
+
+        Args:
+            hide: boolean about hide falsy node or not
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `hiding`

File: backend/core/rule/tree.py [hunk 12]
```diff
@@ -186,40 +352,86 @@ def add_children(self, obj):
         self.children.append(obj)
 
     def eval_children(self, taken_list):
+        """Evaluate children recursively
+
+        Evaluate TreeNode and set self.data to its value.
+        If children exist, evaluated them all recursively.
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `evaluate`

File: backend/core/rule/tree.py [hunk 13]
```diff
@@ -186,40 +352,86 @@ def add_children(self, obj):
...
+            None
+        """
+
+        # First, evaluate all children so that itself can be evaluated directly
         for i, child in enumerate(self.children):
             if isinstance(child, TreeNode):
                 child.eval_children(taken_list)
 
+        # If it is not course, which is not course code number, it should be
+        # evaluated through `func`, because it is not directly able to evaluated
```
Comment (sonowz) [inline backend/core/rule/tree.py]: `able to be evaluated`

File: backend/core/rule/tree.py [hunk 14]
```diff
@@ -186,40 +352,86 @@ def add_children(self, obj):
...
+                    # Change name to what user actually took
+                    self.namespace = course['title']
+
+                    # and consume the entry, because it can cause
+                    # double-evaluating problem, which can lead to
+                    # wrong result.
                     del taken_list[i]
                     break
+
+            # it any of entries of taken_list is not matches,
```
Comment (sonowz) [inline backend/core/rule/tree.py]:
```python
# if any of entries in taken_list does not match,
```
I'm not sure if this logic is right.

**Review by sonowz — CHANGES_REQUESTED** [review]

File: backend/core/rule/tree.py (hunk 1 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: good

File: backend/core/rule/tree.py (hunk 2 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: good

File: backend/core/rule/tree.py (hunk 3 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: good

File: backend/core/rule/tree.py (hunk 4 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: nope, closure is right

closure's meaning can be referenced in [here](https://en.wikipedia.org/wiki/Closure_(computer_programming))

File: backend/core/rule/tree.py (hunk 5 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: good

File: backend/core/rule/tree.py (hunk 6 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: request does not have identical meaning with mine. But, I found one grammatic mistake. I'll fix this.

File: backend/core/rule/tree.py (hunk 7 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: cool

File: backend/core/rule/tree.py (hunk 8 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: well, it is maybe just same meaning, but I will fix it

File: backend/core/rule/tree.py (hunk 9 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: appear is right, but i will change to `be shown`.

File: backend/core/rule/tree.py (hunk 10 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: ok

File: backend/core/rule/tree.py (hunk 11 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: ok

File: backend/core/rule/tree.py (hunk 13 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: cool

File: backend/core/rule/tree.py (hunk 12 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: ok

File: backend/core/rule/tree.py (hunk 14 above)
Comment (skystar-p) [reply backend/core/rule/tree.py]: that's right. I fixed.

**Comment by sonowz** [comment]
Good. I'll merge.

**Review by sonowz — APPROVED** [review]

**Pull request merged**